commit 4bde0f7d1dca0a7d886997eb8dee3fb47a6484e4
Author: Johannes Berg <johannes@sipsolutions.net>
Date:   Tue Jul 7 23:46:51 2009 +0200

    cfg80211: fix two buglets
    
    This fixes two small bugs:
     1) the connect variable is already initialised, and the
        assignment to auth_type overwrites the previous setting
        with a wrong value
     2) when all authentication attempts fail, we need to report
        that we couldn't connect
    
    Signed-off-by: Johannes Berg <johannes@sipsolutions.net>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/net/wireless/sme.c b/net/wireless/sme.c
index df9173f73604..79ca56cbfd36 100644
--- a/net/wireless/sme.c
+++ b/net/wireless/sme.c
@@ -295,9 +295,8 @@ void cfg80211_sme_rx_auth(struct net_device *dev,
 		wdev->conn->state = CFG80211_CONN_AUTHENTICATE_NEXT;
 		schedule_work(&rdev->conn_work);
 	} else if (status_code != WLAN_STATUS_SUCCESS) {
-		wdev->sme_state = CFG80211_SME_IDLE;
-		kfree(wdev->conn);
-		wdev->conn = NULL;
+		__cfg80211_connect_result(dev, mgmt->bssid, NULL, 0, NULL, 0,
+					  status_code, false);
 	} else if (wdev->sme_state == CFG80211_SME_CONNECTING &&
 		 wdev->conn->state == CFG80211_CONN_AUTHENTICATING) {
 		wdev->conn->state = CFG80211_CONN_ASSOCIATE_NEXT;